### PR TITLE
feat: build a row id map during compaction and send it to some kind of index remapper

### PR DESCRIPTION
**Pull request by westonpace**
The actual remapping of the indices will come in a following PR.

### CODE REVIEW

File: rust/lance/src/dataset/optimize.rs [hunk 1]
```diff
@@ -498,15 +554,204 @@ pub struct RewriteResult {
...
+    }
+}
+
+impl datafusion::physical_plan::RecordBatchStream for RowIdCapture {
+    fn schema(&self) -> SchemaRef {
+        self.target.schema()
+    }
+}
+
+struct MissingIds<'a, I: Iterator<Item = u64>> {
```
Comment (wjones127) [inline rust/lance/src/dataset/optimize.rs:634]: not really critical, but: Why make `MissingIds` generic over iterator? It seems like we could likely get some better performance by taking advantage of the unique structure of treemap. For example, we could call [bitmaps](https://docs.rs/roaring/0.10.2/roaring/treemap/struct.RoaringTreemap.html#method.bitmaps) to get an iterator over the fragment_id and local row ids Roaring bitmaps. And then on the local bitmaps, we can go in chunks and call [contains_range](https://docs.rs/roaring/0.10.2/roaring/bitmap/struct.RoaringBitmap.html#method.contains_range) to see if we move on early. That would be optimal for the common case where we have sparse deleted rows (especially given our threshold is only 10%).

**Review by wjones127 — COMMENTED** [review]
A few initial comments.

File: rust/lance/src/dataset/optimize.rs [hunk 2]
```diff
@@ -498,15 +554,204 @@ pub struct RewriteResult {
...
+        let row_ids_arr = batch.column(row_id_idx);
+        let row_ids_itr = row_ids_arr
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .expect(&format!(
+                "Row ids had an unexpected type: {}",
+                row_ids_arr.data_type()
+            ))
+            .iter()
+            .map(|x| x.expect("Null row id encountered"));
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs]: Done.

File: rust/lance/src/dataset/optimize.rs [hunk 3]
```diff
@@ -498,15 +554,204 @@ pub struct RewriteResult {
     pub read_version: u64,
     /// The original fragments being replaced
     pub original_fragments: Vec<Fragment>,
+    pub remapped_indices: Vec<RemappedIndex>,
+}
+
+#[pin_project]
+struct RowIdCapture {
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs]: I switched to this approach.

File: rust/lance/src/dataset/optimize.rs [hunk 4]
```diff
@@ -498,15 +554,204 @@ pub struct RewriteResult {
...
+                Poll::Ready(Some(extracted))
+            }
+            None => Poll::Ready(None),
+        }
+    }
+}
+
+impl datafusion::physical_plan::RecordBatchStream for RowIdCapture {
+    fn schema(&self) -> SchemaRef {
+        self.target.schema()
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs]: Good catch.  Seems I was getting away with it but I fixed this for completeness.

File: rust/lance/src/dataset/optimize.rs (hunk 1 above)
Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs:697]: Fair point.  I didn't even consider this really.  Compaction is already pretty heavily O(# of modified rows) so I figured one more pass through wouldn't hurt.  I'll leave this for a potential perf improvement down the road.

Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs:697]: Added to #1369

File: rust/lance/src/dataset/optimize.rs [hunk 5]
```diff
@@ -85,24 +97,64 @@
...
+
+/// Options for creating an [IndexRemapper]
+pub trait IndexRemapperOptions: std::fmt::Debug + Default + Clone + PartialEq + Serialize {
+    fn create_remapper(&self, dataset: &Dataset) -> Result<Box<dyn IndexRemapper>>;
+}
+
 /// Options to be passed to [compact_files].
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct CompactionOptions {
+pub struct CompactionOptions<I: IndexRemapperOptions = DatasetIndexRemapperOptions> {
```
Comment (westonpace) [inline rust/lance/src/dataset/optimize.rs:157]: My idea here was that "index remapping" should be completely opaque to compaction.  In practice, there are no options that the remapping process needs and there is only one real implementation of this trait.  We could probably implement this in some much simpler way.  Any opinions?

**Review by wjones127 — COMMENTED** [review]
Got a couple questions

File: rust/lance/src/dataset/optimize.rs [hunk 6]
```diff
@@ -29,12 +29,24 @@
 //! 2. If a fragment has a higher percentage of deleted rows than the provided
 //!    threshold.
 //!
+//! In addition to the rules above there may be restrictions due to indexes.
+//! When a fragment is compacted its row ids change and any index that contained
+//! that fragment will be remapped.  There can only be one compaction task per
+//! index and all fragments in that task must be indexed fragments.
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs:35]: I don't have any tests for it yet but the way I believe it works is that you can only compact a set of fragments if those fragments are all covered by the same indices. So, if you have:

frag-1
frag-2
frag-3
frag-4
frag-5
index-1: 1, 2, 3, 4
index-2: 1, 2

Then you can compact 1 & 2 or you can compact 3 & 4.  However, they must be separate compaction runs.  The current task planner should give you 1 & 2 on the first call to plan_compaction and then 3 & 4 on the second call.

File: rust/lance/src/dataset/transaction.rs [hunk 7]
```diff
@@ -402,11 +438,71 @@ impl Transaction {
...
+                    .iter_mut()
+                    .find(|idx| idx.uuid == rewritten_index.old_id)
+                    .ok_or_else(|| {
+                        Error::invalid_input(format!(
+                            "Invalid compaction plan refers to index {} which does not exist",
+                            rewritten_index.old_id
+                        ))
+                    })?;
+
+                dbg!("Old", &index.fragment_bitmap);
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/dataset/transaction.rs:487]: No. :facepalm:

File: rust/lance/src/format/fragment.rs [hunk 8]
```diff
@@ -210,6 +210,54 @@ impl From<&Fragment> for pb::DataFragment {
     }
 }
 
+pub struct RowId(u64);
+
+impl RowId {
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance/src/format/fragment.rs:215]: I like the name `Address`.  I'll update it.

**Comment by westonpace** [comment]
Closing to open up smaller PR into the feature branch

**Pull request closed without merging**